arch/riscv/include/kernel_arch_func.h

```diff
@@ -36,6 +36,20 @@ static ALWAYS_INLINE void arch_kernel_init(void)
 	_kernel.cpus[0].arch.hartid = csr_read(mhartid);
 	_kernel.cpus[0].arch.online = true;
 #endif
+#if ((CONFIG_MP_MAX_NUM_CPUS) > 1)
+	unsigned int cpu_node_list[] = {
+		DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus), DT_REG_ADDR, (,))
+	};
+	unsigned int cpu_num, hart_x;
+
+	for (cpu_num = 1, hart_x = 0; cpu_num < CONFIG_MP_MAX_NUM_CPUS; cpu_num++) {
+		if (cpu_node_list[hart_x] == CONFIG_RV_BOOT_HART) {
```

Comment:
Isn't the point of this code really to skip the currently running 'boot' cpu (a.k.a cpu running this code), which has been setup already above and is guaranteed to have hartid  of  CONFIG_RV_BOOT_HART? Comparing with _kernel.cpus[0].arch.hartid value read from csr_read(mhartid); will spell it more explicitly.
Should the same trick as with cpu_node_list be used to adjust the value returned by arch_num_cpus(). The only other architecture that sometimes returns anything, but CONFIG_MP_MAX_NUM_CPUS is xtensa though.

Reply:
>  Comparing with _kernel.cpus[0].arch.hartid value read from csr_read(mhartid); will spell it more explicitly. Should the same trick as with cpu_node_list be used to adjust the value returned by arch_num_cpus(). The only other architecture that sometimes returns anything, but CONFIG_MP_MAX_NUM_CPUS is xtensa though.

@akabaev, Good spot! pushing a change for this shortly.

>  While there, is running _without_ dts is still a thing supported by Zephyr?

I certainly dont think the riscv_machine_timer would work without dts, as the MTIME register is retrieved from it.

Reply:
@con-pax This was not really a request, just some musings inspired by looking at your changes. For one, I do not have whole lot of the background on the while CONFIG_MP_NUM_CPUS -> CONFIG_MP_MAX_NUM_CPUS and their relationship to  arch_num_cpus() and whether adjustment of  arch_num_cpus() based on DTS content is even desired. My question about running without DTS was more about this: should you be providing fallback cpu_id == mhartid path when there is no DTS to build cpu_node_list from?

Reply:
I guess the point of this PR is to be able to run applications on platforms that have hart id's that are not aligned with Zephyrs sequential cpu numbering scheme. 
The proposed changes here allow for both non-zero indexed and non-contiguous SMP capable hart configurations, thanks to being able to parse the device tree. 
As I see it, the right way to do this is to use the Devicetree. Other implementations may involve more Kconfig options and/or may only allow you to use a contiguous range of hartids. 
The inspiration for this change set was borrowed from what is done in arm64 where they use the same devicetree macros to define a cpu_node_list, which would also fall over if devicetree support was not there.